packages/adminrouter/extra/src/master/cache.lua

```diff
@@ -293,6 +294,82 @@ local function fetch_and_store_state_mesos(auth_token)
 end
 
 
+local function fetch_and_store_mesos_leader_state()
+    -- Lua forbids jumping over local variables definition, hence we define all
+    -- of them here.
+    local res_body, r, err, answers
+
+    if HOST_IP == 'unknown' then
+        ngx.log(ngx.ERR,
+        "Local Mesos Master IP address is unknown, cache entry is unusable")
+        res_body = '{"is_local": "unknown", "leader_ip": null}'
+        goto store_cache
+    end
+
+    -- We want to use Navstar's dual-dispatch approach and get the response
+    -- as fast as possible from any operational MesosDNS. If we set it to local
+    -- instance, its failure will break this part of the cache as well.
+    --
+    -- Also, see https://github.com/openresty/lua-resty-dns#limitations
+    r, err = resolver:new{
+        nameservers = {{"198.51.100.1", 53},
+                       {"198.51.100.2", 53},
+                       {"198.51.100.3", 53}},
+        retrans = 3,  -- retransmissions on receive timeout
+        timeout = 2000,  -- msec
+    }
+
+    if not r then
+        ngx.log(ngx.ERR, "Failed to instantiate the resolver: " .. err)
+        return
+    end
+
+    answers, err = r:query("leader.mesos")
+    if not answers then
+        ngx.log(ngx.ERR, "Failed to query the DNS server: " .. err)
+        return
+    end
+
+    if answers.errcode then
+        ngx.log(ngx.ERR,
+            "DNS server returned error code: " .. answers.errcode .. ": " .. answers.errstr)
+        return
+    end
+
+    if util.table_len(answers) == 0 then
+        ngx.log(ngx.ERR,
+            "DNS server did not return anything for leader.mesos")
+        return
+    end
+
+    -- Yes, we are assuming that leader.mesos will always be just one A entry.
+    -- AAAA support is a different thing...
+
+    if answers[1].address == HOST_IP then
+        res_body = '{"is_local": "yes", "leader_ip": "'.. HOST_IP ..'"}'
+        ngx.log(ngx.INFO, "Mesos Leader is local")
+    else
+        res_body = '{"is_local": "no", "leader_ip": "'.. answers[1].address ..'"}'
+        ngx.log(ngx.INFO, "Mesos Leader is non-local: `" .. answers[1].address .. "`")
+    end
+
+    ::store_cache::
+    if not cache_data("mesos_leader", res_body) then
```

Comment:
Maybe we can add some comment that we're on purpose ignoring `TTL` of the record and going with own refresh cycle period. I am also wondering what should we do in case if record TTL is shorter than our cache period.